### DISCUSSION TITLE
[JS] 제너레이터

### DISCUSSION THREAD

**Discussion by WooGie911** [wanted-9th-3team/frontend-study#75] [프론트 기술 질문 공간]
제너레이터에 대해 설명해주세요.

**Comment by michoball** [comment]
제너레이터 함수는 함수 안의 코드를 이터레이트 하기위한 방법입니다. 
제너레이터 함수는 숨겨진 Iterator class 의 하위 클래스로 반복가능한 제너레이터 객체를 반환합니다. 

function 뒤에 * 를 붙여서 생성하고 함수 내부에 yield 키워드를 사용하여 해당 부분에서 함수의 실행을 중지시킬 수 있습니다. 

제너레이터 함수가 반환한 제너레이터 객체에서 next 메서드를 사용해서 함수한 yield 문까지 함수를 끊어서 진행할 수 있습니다. 

함수의 모든  yield를 지나거나 return 에 도달하게 되면 제너레이터 객체의 done 값는 true가 되고 이후의 value 값은 undefined가 됩니다. 

반복가능한 함수이기 때문에 for - of 문을 사용할 수 있습니다. 
mdn 예제
```javascript
const foo = function*() {
  yield 'a';
  yield 'b';
  yield 'c';
};

let str = '';
for (const val of foo()) {
  str = str + val;
}

console.log(str);
// Expected output: "abc"

```

**Comment by WooGie911** [comment]
일반 함수는 하나의 값(혹은 0개의 값)만을 반환합니다.
하지만 제너레이터(generator)를 사용하면 여러 개의 값을 필요에 따라 하나씩 반환(yield)할 수 있습니다. 
다시말해 제너레이터 함수는 중간에 원하는 부분에서 멈추었다가 그 부분부터 다시 실행할수있는 능력을 가진 함수입니다.

제너레이터와 이터러블(iterable, 반복 가능한) 객체를 함께 사용하면 손쉽게 데이터 스트림을 만들 수 있으며,
제너레이터를 만들려면 '제너레이터 함수’라 불리는 특별한 문법 구조, function*이 필요합니다.

제너레이터 함수는 일반 함수와 동작 방식이 다릅니다. 
제너레이터 함수를 호출하면 코드가 실행되지 않고, 대신 실행을 처리하는 특별 객체, '제너레이터 객체’가 반환됩니다.

정리하면,
- 제너레이터는 제너레이터 함수 function* f(…) {…}을 사용해 만듭니다.
- yield 연산자는 제너레이터 안에 있어야 합니다.
- next/yield 호출을 사용하면 외부 코드와 제너레이터 간에 결과를 교환할 수 있습니다.
